### ISSUE TITLE
[api-extractor] Linking to overloads generates warnings, but no way to specify which overload

### ISSUE THREAD

**Issue by AlexJerabek** [microsoft/rushstack#1240]
Our d.ts file has two namespaces named "Office". When running API Extractor version 7.1.0, links like this `{@link Office.AsyncResult | AsyncResult}` give warnings like this:
> Warning: office.d.ts:2842:9 - (ae-unresolved-link) The @link reference could not be resolved: The reference is ambiguous because "Office" has more than one declaration; you need to add a TSDoc member reference selector

I wanted to use the guidance given  in [this TSDoc spec](https://github.com/Microsoft/tsdoc/blob/master/spec/code-snippets/DeclarationReferences.ts), but I'm unable to use the labels or index selectors to reference overloaded namespaces. 

When using `{@link (Office:2).AsyncResult}`, I get the warning:
> Warning: office.d.ts:328:5 - (ae-unresolved-link) The @link reference could not be resolved: The selector "2" is not a supported selector type

When using `{@label COMMON_API}`, I get the warning:
> Warning: office.d.ts:7:5 - (tsdoc-unsupported-tag) The TSDoc tag "@label" is not supported by this tool

Is there a workaround until labels or index-references are supported? Thank you.

**Comment by octogonz** [comment]
Distinguishing declarations is generally useful for merged declarations (e.g. overloaded functions, enum+namespace, interface+function, etc), but for namespace+namespace or interface+interface are special cases where the distinction between declarations isn't meaningful.

If you have two namespaces named "Office", I suppose ideally API Extractor should collapse them into a single namespace.  However this can involve nontrivial transformations, for example: 

```ts
// "Oh, TypeScript!"  :-D
export namespace N {
  export interface A { }
  export interface C extends B { }
}

export namespace N  {
  export interface B { }
  export interface C extends A { }
}
```

The collapsed namespace would need a collapsed `export interface C extends A, B { }`.

For the special case of namespace+namespace (and interface+interface), I'm tempted to propose that the declaration reference resolver should arbitrarily select the first one it finds.  In practice it doesn't matter much.  This would eliminate the error and need for `@label`.

But @AlexJerabek... now I'm wondering how these multiple namespaces appear in your generated docs.  Do they get folded together?  Do they get separate TOC nodes?  Does one of them get silently discarded?  If the answer is something bad, then maybe we should go ahead and implement collapsing, at least in the .api.json file if not the .d.ts rollup.

(We also absolutely need to implement support for `@label` and `:1`.  This isn't a lot of work.  It was procrastinated merely because it didn't meet AE7's release bar of achieving feature parity with AE6.  I don't believe it's needed for this case, but if you need it for overloaded functions, please open a separate GitHub issue to track it.)

**Comment by AlexJerabek** [comment]
Thanks for the quick response @octogonz!

The namespaces get folded together today, which is what we'd expect. We can merge the namespaces together in our d.ts (apparently the separation was for historical reasons). That avoids this issue.

Because we're not blocked, this isn't an API Extractor feature we need. This issue was more to point out the discrepancy between the warning's advice and actual functionality.

**Comment by octogonz** [comment]
The short answer is that it's not implemented.  (Fixing this is obviously a high priority in general, but embarrassingly it's stagnated because it doesn't affect us much; the Rush Stack project tries to avoid designing APIs with overloads.)

There are two versions of the `@link` syntax:
- OLD: `{@link @thi.ng/arrays#(quickSort:1)}`
- NEW: `{@link @thi.ng/arrays!quickSort:function(1)}`

The new notation was introduced this summer and has not been integrated yet into the TSDoc parser library. That work needs to be completed before API Extractor could adopt it (even though API Extractor is already using it for the `"canonicalReference"` ID).

Thus, if we want to get this working quickly, we should implement it for the old syntax.  It's probably not a lot of work, and much of the implementation should be reusable later when we switch to the new syntax.

If I remember right, this needs to be implemented in two places:
- [AstReferenceResolver._selectDeclaration()](https://github.com/microsoft/rushstack/blob/bdc77711a5546f7f4681f72a82d4a5769719e4e0/apps/api-extractor/src/analyzer/AstReferenceResolver.ts#L129) - what API Extractor uses during analysis
- [ModelReferenceResolver.resolve()](https://github.com/microsoft/rushstack/blob/bdc77711a5546f7f4681f72a82d4a5769719e4e0/apps/api-extractor-model/src/model/ModelReferenceResolver.ts#L112) - what API Documenter uses when working with .api.json files

Would someone want to help out with this?  I'm pretty busy right now, so I'm not sure I could do it right away.  But I'd be very happy to provide guidance/help if someone else wants to make the PR.

**Comment by octogonz** [comment]
Here's a PR: https://github.com/microsoft/rushstack/pull/1648

@postspectacular  @AlexJerabek

**Comment by octogonz** [comment]
This fix was published with API Extractor 7.7.0

**Comment by postspectacular** [comment]
@octogonz you're a ⭐️ ! thank you so much & apols for lack of responses. busy times!
